rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,373 @@
+- Kibana calculates whether a user should receive a notification in runtime before showing one to the customer.
+  The only available option for the current Kibana Security model. Considering the future integration with UNS,
+  Kibana has to provide an HTTP API to check users' permissions are sufficient to see a notification.
+  Unified Notification center performs a call to HTTP API to filter out notifications before rendering them in Cloud UI. 
+  This option adds a significant runtime performance penalty. Moreover, it makes the rendering process dependent on Kibana instance availability. If one of them is temporarily unavailable, the UNS can't decide if a notification must be shown or not.
+
+### Notification System
+Abstracts the way notifications are created, stored, configured, and retrieved.
+It allows Kibana to adopt different strategies for on-premise and Cloud versions:
+- *LocalRepository* - a repository for on-premise version, it stores data within the current deployment. 
+- *RemoteRepository* - a repository for the Cloud version of Kibana. It stores data within the Unified Notification System deployed in the Cloud.
+In this case, a user can see in Kibana UI the external notifications created outside of Kibana.
+
+*Repository* interface:
+- `create(NotificationEvent)`- creates a notification in the Notification system. 
+- `get({ search_after?: number; size?: number; type_id?: string[] })` - retrieves last notifications for the current user.
+Might be paginated using `search_after` filed.
+- `markAsRead({ notification_id: string[] })` - mark notifications with given ids as read.
+- `markAsUnread({ notification_id: string[] })` - mark notifications with given ids as unread.
+- `markAllAsRead()` - mark all the notifications as read.
+- `markAsPin({ notification_id: string[] })` - mark notifications with given ids as pinned.
+- `markAsUnpin({ notification_id: string[] })` - mark notifications with given ids as unpinned.
+- `configure()` - updates user-specific notification settings.
+
+#### LocalRepository
+![image](../images/0014/local_repository.png)
+
+##### Notification model
+Notification model is responsible for:
+- Applying policies to an incoming notification. NS must check whether a recipient unsubscribed from a source before processing to delivery.
+- Enhancing an incoming notification with necessary data: 
+    - `expired_at` if not specified. Kibana allows administrators to set how long notifications will be stored. Falls back to 30 - 90 days be default. 
+- Writing new incoming messages to the Notification storage.
+
+Processing a large number of incoming notifications might slow down Kibana and create a disproportionately high load on some Kibana instances.
+The use of a queue allows us to make the handling of incoming notifications asynchronously.
+Async mechanism lets Kibana to cope with spikes in the number of messages created. The queue implementation is based on the Elasticsearch server. 
```

Comment:
> Is it possible that we could implement an MVP without this queue? How do we evaluate when we need this? Just concerned this may be a premature optimization.

We shouldn't implement it from the very beginning. In the `Adoption strategy` section, `Phase III`: `Load testing is done. Make adjustments according to the number of unique users: notification queue`.